File: windows/src/ConnectorControl.Core/State/SettingsModel.cs
using ConnectorControl.Core.Services;

namespace ConnectorControl.Core.State;

/// <summary>Catalog §4 SettingsView state: three tabs, every toggle, path, and button.</summary>
public sealed class SettingsModel : ObservableObject
{
    public const string GeneralTab = "General";
    public const string StorageTab = "Storage";
    public const string ClaudeTab = "Claude";
    public const string LaunchAtStartupTitle = "Launch at startup";
    public const string ConfirmRestartTitle = "Confirm before restarting Claude";
    public const string ConfirmQuitTitle = "Confirm before quitting";
    public const string NotifyTitle = "Notify about changes made outside Connector Control";
    public const string NotifyCaption = "Covers edits to Claude's config and synced connector-list changes, including when a remote change needs a Claude restart.";
    public const string UpdatesHeader = "Updates";
    public const string AutoUpdateTitle = "Automatically download and install updates";
    public const string CheckForUpdatesTitle = "Check for Updates…";
    public const string MasterListHeader = "Master List Location";
    public const string ChooseTitle = "Choose…";
    public const string UseDefaultTitle = "Use Default";
    public const string BackupsHeader = "Backups";
    public const string BackupsCaption = "Both config files are backed up automatically before every change.";
    public const string ShowInExplorerTitle = "Show in Explorer";
    public const string RestoreTitle = "Restore…";
    public const string ClaudeAppHeader = "Claude App";
    public const string ConfigPathLabel = "Config file";
    public const string LaunchTargetLabel = "Launch target";
    public const string NotFoundText = "Not found";
    public const int MinKeepCount = 5;
    public const int MaxKeepCount = 100;

    private readonly AppState state;
    private readonly ISettings settings;
    private readonly IAutostart autostart;
    private readonly IClaudeInstall install;
    private readonly IUpdater updater;
    private readonly UpdateCoordinator updates;
    private bool launchAtStartup;
    private string? loginItemNote;

    public SettingsModel(AppState state, ISettings settings, IAutostart autostart, IClaudeInstall install, IUpdater updater, UpdateCoordinator updates)
    {
        this.state = state;
        this.settings = settings;
        this.autostart = autostart;
        this.install = install;
        this.updater = updater;
        this.updates = updates;
        launchAtStartup = autostart.IsEnabled;
    }

    /// <summary>Called whenever the window opens: autostart is read fresh (the user may have changed it in Windows Settings).</summary>
    public void Refresh()
    {
        launchAtStartup = autostart.IsEnabled;
        RaiseAll();
    }

    // MARK: General (catalog §4.2)

    /// <summary>Catalog §4.2: no-op when the OS already agrees; on failure revert the toggle and show the note.</summary>
    public bool LaunchAtStartup
    {
        get => launchAtStartup;
        set
        {
            if (!Set(ref launchAtStartup, value))
            {
                return;
            }
            var actual = autostart.IsEnabled;
            if (value == actual)
            {
                return;
            }
            try
            {
                autostart.SetEnabled(value);
                LoginItemNote = null;
            }
            catch (InvalidOperationException ex)
            {
                launchAtStartup = actual;
                Raise(nameof(LaunchAtStartup));
                LoginItemNote = $"Couldn't update login item: {ex.Message}";
            }
        }
    }

    public string? LoginItemNote { get => loginItemNote; private set => Set(ref loginItemNote, value); }

    public bool ConfirmBeforeRestart
    {
        get => settings.ConfirmBeforeRestart;
        set { settings.ConfirmBeforeRestart = value; Raise(); }
    }

    public bool ConfirmBeforeQuit
    {
        get => settings.ConfirmBeforeQuit;
        set { settings.ConfirmBeforeQuit = value; Raise(); }
    }

    public bool NotifyExternalChanges
    {
        get => settings.NotifyExternalChanges;
        set { settings.NotifyExternalChanges = value; Raise(); }
    }

    public bool AutoUpdate
    {
        get => settings.AutoUpdate;
        set { settings.AutoUpdate = value; Raise(); }
    }

    public bool UpdatesEnabled => updater.IsAvailable;

    public string VersionText => $"Version {updater.VersionDisplay}";

    public Task CheckForUpdatesAsync() => updates.CheckAsync(interactive: true);

    // MARK: Storage (catalog §4.3)

    public string StoreDirPath => state.Service.Paths.StoreDir;

    public bool CanUseDefaultStore => !string.IsNullOrEmpty(settings.MasterStoreDir);

    public void ChooseStoreDir(string dir)
    {
        state.RepointStore(dir);
        RaiseAll();
    }

    public void UseDefaultStoreDir()
    {
        state.RepointStore(null);
        RaiseAll();
    }

    public int BackupKeepCount
    {
        get => settings.BackupKeepCount;
        set
        {
            var clamped = Math.Clamp(value, MinKeepCount, MaxKeepCount);
            if (clamped == settings.BackupKeepCount)
            {
                Raise();                        // the stepper's own binding snaps back to the clamped value
                Raise(nameof(KeepCountLabel));  // same two raises on both branches, so the label can never drift
                return;
            }
            settings.BackupKeepCount = clamped;
            state.RefreshServiceSettings();
            Raise();
            Raise(nameof(KeepCountLabel));
        }
    }

    public string KeepCountLabel => $"Keep {settings.BackupKeepCount} backups of each file";

    public void IncrementKeepCount() => BackupKeepCount = settings.BackupKeepCount + 1;

    public void DecrementKeepCount() => BackupKeepCount = settings.BackupKeepCount - 1;

    public string BackupsDir => state.Service.Backups.BackupsDir;

    // MARK: Claude (catalog §4.4, spec §7.3)

    public string InstallKindText => install.Detect().Kind switch
    {
        ClaudeInstallKind.Msix => "MSIX package",
        ClaudeInstallKind.Legacy => "Legacy installer",
        _ => NotFoundText,
    };

    public string ClaudeConfigPath => state.Service.Paths.ClaudeConfigPath;

    public bool CanUseDefaultClaudeConfig => !string.IsNullOrEmpty(settings.ClaudeConfigPath);

    public void ChooseClaudeConfig(string path)
    {
        state.RepointClaudeConfig(path);
        RaiseAll();
    }

    public void UseDefaultClaudeConfig()
    {
        state.RepointClaudeConfig(null);
        RaiseAll();
    }

    public string LaunchTargetText =>
        settings.ClaudeLaunchTarget is { Length: > 0 } overridden ? overridden : install.Detect().LaunchTarget ?? NotFoundText;

    public bool CanUseDefaultLaunchTarget => !string.IsNullOrEmpty(settings.ClaudeLaunchTarget);

    public void ChooseLaunchTarget(string exe)
    {
        settings.ClaudeLaunchTarget = exe;
        RaiseAll();
    }

    public void UseDefaultLaunchTarget()
    {
        settings.ClaudeLaunchTarget = null;
        RaiseAll();
    }
}

File: Sources/ConnectorControlState/SettingsModel.swift
import Foundation
import Combine
import ConnectorControlCore

/// Catalog §4 SettingsView state: three tabs, every toggle, path, and button.
/// Pass-through settings are computed properties that write the seam and send
/// objectWillChange (the C# Raise()).
@MainActor
public final class SettingsModel: ObservableObject {
    public static let generalTab = "General"
    public static let storageTab = "Storage"
    public static let claudeTab = "Claude"
    public static let launchAtLoginTitle = "Launch at login"
    public static let confirmRestartTitle = "Confirm before restarting Claude"
    public static let confirmQuitTitle = "Confirm before quitting"
    public static let notifyTitle = "Notify about changes made outside Connector Control"
    public static let notifyCaption = "Covers edits to Claude's config and synced connector-list changes, including when a remote change needs a Claude restart."
    public static let updatesHeader = "Updates"
    public static let autoUpdateTitle = "Automatically download and install updates"
    public static let checkForUpdatesTitle = "Check for Updates…"
    public static let masterListHeader = "Master List Location"
    public static let chooseTitle = "Choose…"
    public static let useDefaultTitle = "Use Default"
    public static let backupsHeader = "Backups"
    public static let backupsCaption = "Both config files are backed up automatically before every change."
    public static let revealInFinderTitle = "Reveal in Finder"
    public static let restoreTitle = "Restore…"
    public static let claudeAppHeader = "Claude App"
    public static let claudeAppRejectedTitle = "That app can’t be used as Claude Desktop"
    public static let toolsHeader = ToolNote.settingsHeader
    public static let toolsCaption = ToolNote.settingsCaption
    public static let loginItemApprovalNote = "Approve Connector Control under System Settings → General → Login Items."
    public static let minKeepCount = 5
    public static let maxKeepCount = 100

    public static func versionText(_ version: String) -> String { "Version \(version)" }

    public static func keepCountLabel(_ count: Int) -> String { "Keep \(count) backups of each file" }

    public static func loginItemFailureNote(_ message: String) -> String { "Couldn't update login item: \(message)" }

    private let state: AppState
    private let settings: AppSettings
    private let autostart: Autostart
    private let updater: Updater
    private var subscriptions: Set<AnyCancellable> = []
    private var settingLaunchAtLoginSilently = false

    public init(state: AppState, settings: AppSettings, autostart: Autostart, updater: Updater) {
        self.state = state
        self.settings = settings
        self.autostart = autostart
        self.updater = updater
        launchAtLogin = autostart.isEnabled
        // Only the tool statuses are relayed from AppState. The Storage tab's
        // path and keep count derive from state.service, whose only writers
        // (repointStore, refreshServiceSettings) are called from this model,
        // which raises objectWillChange itself first. An AppState-side repoint
        // added later would need a relay of state.objectWillChange here.
        state.$toolStatuses.dropFirst()
            .sink { [weak self] _ in self?.objectWillChange.send() }
            .store(in: &subscriptions)
        // A change made in Sparkle's own dialog must not leave the toggle stale.
        updater.automaticallyDownloadsUpdatesPublisher.dropFirst()
            .sink { [weak self] _ in self?.objectWillChange.send() }
            .store(in: &subscriptions)
    }

    /// Called whenever the General tab appears: autostart is read fresh (the
    /// user may have changed it in System Settings), and the updater's flag re-read.
    public func refresh() {
        setLaunchAtLoginSilently(autostart.isEnabled)
        objectWillChange.send()
    }

    // MARK: - General (catalog §4.2)

    /// Catalog §4.2: no-op when the OS already agrees; on failure revert the
    /// toggle and show the note; when approval is pending, say where.
    @Published public var launchAtLogin: Bool {
        didSet {
            if !settingLaunchAtLoginSilently, oldValue != launchAtLogin {
                applyLaunchAtLogin(launchAtLogin)
            }
        }
    }

    @Published public private(set) var loginItemNote: String?

    private func setLaunchAtLoginSilently(_ value: Bool) {
        settingLaunchAtLoginSilently = true
        launchAtLogin = value
        settingLaunchAtLoginSilently = false
    }

    private func applyLaunchAtLogin(_ wantOn: Bool) {
        let isOn = autostart.isEnabled
        guard wantOn != isOn else { return }
        do {
            try autostart.setEnabled(wantOn)
            loginItemNote = nil
        } catch {
            setLaunchAtLoginSilently(isOn)   // revert the checkbox
            loginItemNote = SettingsModel.loginItemFailureNote(error.localizedDescription)
        }
        if wantOn, autostart.requiresApproval {
            loginItemNote = SettingsModel.loginItemApprovalNote
        }
    }

    public var confirmBeforeRestart: Bool {
        get { settings.confirmBeforeRestart }
        set {
            objectWillChange.send()
            settings.confirmBeforeRestart = newValue
        }
    }

    public var confirmBeforeQuit: Bool {
        get { settings.confirmBeforeQuit }
        set {
            objectWillChange.send()
            settings.confirmBeforeQuit = newValue
        }
    }

    public var notifyExternalChanges: Bool {
        get { settings.notifyExternalChanges }
        set {
            objectWillChange.send()
            settings.notifyExternalChanges = newValue
        }
    }

    /// Mirrors Sparkle's own flag (it persists it itself).
    public var autoUpdate: Bool {
        get { updater.automaticallyDownloadsUpdates }
        set {
            guard updater.automaticallyDownloadsUpdates != newValue else { return }
            objectWillChange.send()
            updater.automaticallyDownloadsUpdates = newValue
        }
    }

    public var updatesEnabled: Bool { updater.isAvailable }

    public var versionText: String { SettingsModel.versionText(updater.versionDisplay) }

    public func checkForUpdates() { updater.checkForUpdates() }

    // MARK: - Storage (catalog §4.3)

    public var storeDirPath: String { state.service.paths.storeDirURL.path }

    public var canUseDefaultStore: Bool { !(settings.masterStoreDir ?? "").isEmpty }

    public func chooseStoreDir(_ dir: URL) {
        objectWillChange.send()
        state.repointStore(to: dir)
    }

    public func useDefaultStoreDir() {
        objectWillChange.send()
        state.repointStore(to: nil)
    }

    public var backupKeepCount: Int {
        get { settings.backupKeepCount }
        set {
            let clamped = min(max(newValue, SettingsModel.minKeepCount), SettingsModel.maxKeepCount)
            objectWillChange.send()   // the stepper's own binding snaps back to the clamped value
            guard clamped != settings.backupKeepCount else { return }
            settings.backupKeepCount = clamped
            state.refreshServiceSettings()
        }
    }

    public var keepCountLabel: String { SettingsModel.keepCountLabel(settings.backupKeepCount) }

    public func incrementKeepCount() { backupKeepCount = settings.backupKeepCount + 1 }

    public func decrementKeepCount() { backupKeepCount = settings.backupKeepCount - 1 }

    public var backupsDir: URL { state.service.backups.backupsDir }

    // MARK: - Claude (catalog §4.4)

    public var claudeAppPath: String { settings.claudeAppPath ?? AppState.defaultClaudeAppPath }

    public var canUseDefaultClaudeApp: Bool { claudeAppPath != AppState.defaultClaudeAppPath }

    public func chooseClaudeApp(_ app: URL) {
        objectWillChange.send()
        settings.claudeAppPath = app.path
    }

    public func useDefaultClaudeApp() {
        objectWillChange.send()
        settings.claudeAppPath = nil
    }

    // MARK: - Tools (spec 2026-09-05-tool-probe §3.5)

    public var toolRows: [ToolRow] {
        Tool.allCases.map { ToolRow.make(tool: $0, status: state.toolStatuses[$0]) }
    }

    /// Spec §6 D4: the Mac probes all four when the Claude tab appears.
    public func refreshTools() { state.refreshTools() }

    /// Stops listening to AppState and the updater. The app does not call
    /// this: the subscriptions hold `self` weakly and die with the
    /// `@StateObject`. Tests call it to prove the relays are what repaint
    /// the view.
    public func dispose() { subscriptions.removeAll() }
}
